### DISCUSSION TITLE
Terminating data writing when using repeat

### DISCUSSION THREAD

**Discussion by inchinn1** [bonsai-rx/bonsai#1582] [Q&A]
I am trying to record a closed-loop experiment using a teensy 4.1 and a camera. The teensy records 2 analog inputs, corresponding. to when a stimulus valve opens and the output of a treadmill. The teensy also receives a digital input from the stimulator to trigger the recording of a stimulus trial. 
The goal is to record video and the analog inputs for 10 s of baseline before the stimulus and for 25 s after the start of the stimulus. I have a condition that the stimulus is only given if the treadmill has not moved over a recent time window, otherwise, the stimulus is delayed until this condition is met (and the total recording will be extended appropriately).
![bda33f64-bdc1-4c10-a18a-786beda107a5](https://github.com/bonsai-rx/bonsai/assets/21177232/7c9d7235-c2f7-43db-9843-c454cf13dc94)
The red box in the workflow indicates how I have approached the timing of the stop signal for all the 'TakeUntil' nodes.  The digital output is used to signal to the stimulator whether the condition is met.  The first delay started by the trigger is the 10s of baseline, and the second delay is 25s from the stimulus being delivered, after which point I want to terminate recordings.
  
I have multiple trials to record and have used Repeat nodes to increment the recording of each trial.

The workflow performs as expected for the first trial, ≥35s of data are recorded.  However, although subsequent trials start as expected with each trigger, they immediately stop recording.  It seems that the stop signal is emitted without the included delays on every trial after the first. I have tried a Repeat node at the end of the red box but get the same behavior. Is there a way to reset the red box with each trigger? or is there a more elegant approach to what I am trying to achieve?

The complete workflow is also attached.

Very appreciative of all the work that has gone in to Bonsai and any help offered with this problem

[closedloop.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/12805893/closedloop.bonsai.zip)

**Comment by ncguilbeault** [comment]
You have the right idea. Basically what you want is to set up a series of stages or states that your experiment will transition between. If I'm interpreting it correctly, you want something like this.

1) Wait for the trigger to start the experiment.
2) Begin baseline recording. Wait for a minimum of 10 s.
3) Check the state of the treadmill. Wait until the treadmill has stopped moving.
4) Turn on the stimulus. Wait for 25 s.
5) Turn off the stimulus and stop recording. Once completed, go back to stage 1.

With this in mind, you can build your workflow to move between each of these stages using the `SelectMany` module. For example:
![Screenshot from 2023-10-04 16-13-16](https://github.com/bonsai-rx/bonsai/assets/26148508/3c0bb21e-59c7-4ddf-b68d-728bc1437ebe)

The `SelectMany` node lets us enter into a discrete state which executes some logic before sending notifications to downstream nodes. Since the action of `trigger` is to start recording, the `BaselineRecording` group is just a timer that waits for a fixed period. Once this period ends, the `ExtendedRecording` group starts, which tests whether or not the treadmill has been active. If the treadmill is less than a certain value over a certain time window, then it triggers the `StimulusOn` state, which then enters into a `FixedRecordingPeriod` of 25 s. Once that is finished, the experiment stop signal is triggered, followed by `StimulusOff`. Finally `Take` is used to complete the sequence and `Repeat` is used to start the sequence again.

Hopefully this helps!
[closedloop.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/12807898/closedloop.bonsai.zip)

**Reply by inchinn1** [reply]
Great, I have it working! thanks a lot for your help.
